tony-history-server/app/utils/ParserUtils.java

```diff
@@ -144,5 +160,55 @@ public static JobMetadata parseMetadata(FileSystem fs, Path jobFolderPath, Strin
     return configs;
   }
 
+  /**
+   * Assuming that there's only 1 jhist file in {@code jobFolderPath},
+   * this function parses the jhist and return a list of {@code JobEvent} objects.
+   * @param fs FileSystem object.
+   * @param jobFolderPath Path object of job directory.
+   * @return a list of {@code JobEvent} objects.
+   */
+  public static List<JobEvent> parseEvents(FileSystem fs, Path jobFolderPath) {
+    if (!pathExists(fs, jobFolderPath)) {
+      return Collections.emptyList();
+    }
+
+    String jhistFile = getJhistFileName(fs, jobFolderPath);
+    if (jhistFile.isEmpty()) {
+      return Collections.emptyList();
+    }
+
+    Path historyFile = new Path(jhistFile);
+    InputStream in = null;
+    try {
+      in = fs.open(historyFile);
+    } catch (IOException e) {
+      LOG.error("Failed to open history file", e);
+    }
+
+    DatumReader<Event> datumReader = new SpecificDatumReader<>(Event.class);
+    List<JobEvent> events = new ArrayList<>();
+    try (DataFileStream<Event> avroFileStream = new DataFileStream<>(in, datumReader)) {
+      Event record;
+      while (avroFileStream.hasNext()) {
+        record = avroFileStream.next();
```

Comment:
I think you can use the `.next(T) ` method call to reuse the existing `Event` object: https://avro.apache.org/docs/1.7.7/api/java/org/apache/avro/file/DataFileStream.html#next(D)

I think you'll to initialize to a dummy `Event` on line 191, though.

Reply:
Yeah I saw that too but I thought that was kinda weird so I went with this instead.

Reply:
We'd have to benchmark it to be certain, but I think reusing the same `Event` object would be more performant, since it saves a lot of unnecessary object allocations and garbage collection.

Reply:
Fixed.